### ISSUE TITLE
New rule proposal: `no-invalid-new-target`

### ISSUE THREAD

**Issue by g-plane** [eslint/eslint#14302]
**Please describe what the rule should do:**

Enforce correct usage of `new.target`.

**What new ECMAScript feature does this rule relate to?**

ES2015 feature: `new.target`.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
new.target;  // out of a function or class

class C {
  method() {
    new.target;  // should be in constructor of a class
  }
}

const f2 = () => new.target;  // can't be in an arrow function
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It's related to ECMAScript language feature, and it's not limited in a specific domain.

**Are you willing to submit a pull request to implement this rule?**

Yes.

**Comment by ljharb** [comment]
Note that it can be in an arrow function defined inside a constructor, and it can be in a normal function anywhere, but not an async function or a generator or an async generator or a concise object method.

**Comment by aladdin-add** [comment]
wondering how much useful the rule will be - I've never seen someone written codes like the examples. ;)

another background: some of these was causing a parsing error. e.g.
![image](https://user-images.githubusercontent.com/13050025/114061257-cfac7880-98c8-11eb-8e3e-6606245b2979.png)

**Comment by nzakas** [comment]
I’m also not sure how useful this would be. If the parser is already catching most illegal uses, I’m not sure a new rule makes sense.
